File: test/scripts/lock-contention.py
#!/usr/bin/env python3
"""
ANTI-PATTERN: Heavy lock contention
Resume: "Reduce Context Switching", "Minimize Lock Contention" - Many
threads competing for one lock causes blocking and context switches.

Run: python lock-contention.py
Measure: pidstat -w
"""

import threading
import time

NUM_THREADS = 32
ITERATIONS = 50_000

lock = threading.Lock()
counter = 0


def increment():
    global counter
    for _ in range(ITERATIONS):
        with lock:  # All threads fight for same lock
            counter += 1


def main():
    print("ANTI-PATTERN: Lock contention")
    print(f"{NUM_THREADS} threads, {ITERATIONS} increments each, single lock")

    start = time.perf_counter()
    threads = [threading.Thread(target=increment) for _ in range(NUM_THREADS)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    elapsed = time.perf_counter() - start

    print(f"With lock: {elapsed*1000:.0f} ms, counter={counter}")
    print("Fix: use ThreadLocal, lock-free (atomic), or reduce critical section")


if __name__ == "__main__":
    main()

File: test/scripts/lock-contention.js
#!/usr/bin/env node
/**
 * ANTI-PATTERN: Heavy lock contention (Worker threads)
 * Resume: "Minimize Lock Contention" - Many workers competing for one lock.
 *
 * Run: node lock-contention.js
 */

const { Worker, isMainThread, workerData } = require('worker_threads');

const NUM_WORKERS = 8;
const ITERATIONS = 50_000;

if (isMainThread) {
  const sab = new SharedArrayBuffer(8);
  const arr = new Int32Array(sab);

  console.log('ANTI-PATTERN: Lock contention');
  console.log(`${NUM_WORKERS} workers, ${ITERATIONS} increments each`);

  const start = Date.now();
  const workers = [];
  for (let i = 0; i < NUM_WORKERS; i++) {
    workers.push(new Promise((resolve) => {
      const w = new Worker(__filename, { workerData: { sab } });
      w.on('exit', resolve);
    }));
  }

  Promise.all(workers).then(() => {
    const elapsed = (Date.now() - start) / 1000;
    console.log(`Done in ${(elapsed * 1000).toFixed(0)} ms, counter=${arr[1]}`);
    console.log('Fix: use lock-free (Atomics.add) or per-worker counters');
  });
} else {
  const { sab } = workerData;
  const arr = new Int32Array(sab);
  for (let i = 0; i < ITERATIONS; i++) {
    while (Atomics.compareExchange(arr, 0, 0, 1) !== 0) {}
    arr[1]++;
    Atomics.store(arr, 0, 0);
  }
}
